pkg/cloud/services/networking/port.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package networking
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/portsbinding"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/portsecurity"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/cluster-api/util"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha4"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/record"
+	capoerrors "sigs.k8s.io/cluster-api-provider-openstack/pkg/utils/errors"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/utils/names"
+)
+
+const (
+	timeoutPortDelete       = 3 * time.Minute
+	retryIntervalPortDelete = 5 * time.Second
+)
+
+func (s *Service) getPort(portID string) (port *ports.Port, err error) {
+	if portID == "" {
+		return nil, fmt.Errorf("portID should be specified to get detail")
+	}
+	port, err = s.client.GetPort(portID)
+	if err != nil {
+		if capoerrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("get port %q detail failed: %v", portID, err)
+	}
+	return port, nil
+}
+
+// GetPortFromInstanceIP returns at most one port attached to the instance with given ID
+// and with the IP address provided.
+func (s *Service) GetPortFromInstanceIP(instanceID string, ip string) ([]ports.Port, error) {
```

Comment:
This is the only function I had to create (to avoid a circular dependency with `GetManagementPort`). All other functions in this file are moved from instance.go.